Apsitvarkom.DataAccess/GoogleGeocoder.cs

```diff
@@ -0,0 +1,42 @@
+﻿using Apsitvarkom.Configuration;
+using Apsitvarkom.Models;
+using System.Text.Json;
+using Yoh.Text.Json.NamingPolicies;
+
+namespace Apsitvarkom.DataAccess;
+
+public class GoogleGeocoder : IGeocoder
+{
+    private readonly HttpClient _httpClient;
+    private readonly string _apiKey;
+
+    private static readonly JsonSerializerOptions SerializerOptions = new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicies.SnakeCaseLower
+    };
+
+    public GoogleGeocoder(HttpClient httpClient, IApiKeyProvider apiKeyProvider)
+    {
+        _apiKey = apiKeyProvider.Geocoding;
+        _httpClient = httpClient;
+    }
+
+    public async Task<string?> ReverseGeocodeAsync(Coordinates coordinates)
+    {
+        var jsonStream = await _httpClient.GetStreamAsync($"json?latlng={coordinates.Latitude},{coordinates.Longitude}&language=lt&result_type=political&key={_apiKey}");
+        var response = await JsonSerializer.DeserializeAsync<ReverseGeocodingApiResponse>(jsonStream, SerializerOptions);
+
+        var title = response?.Results?.FirstOrDefault()?.FormattedAddress;
+        return title ?? string.Empty;
```

Comment:
Should we return an empty string in case we can't resolve the location? Maybe 'Unknown location' would be better for in-web usage, as the title would still exist, instead of it being simply blank?

Reply:
Yeah good comment, in fact we have the method declaration to return `string?` and currently we can't return `null` in any way.

Your suggestion makes sense at first, but here's why we shouldn't do that when you think about it: by adding a default value such as "Unknown location" we're assigning a value to a field that will probably get rendered by our API users (for example our front end). This poses a problem since we're putting it in a specific language. So if our front end has `lt` language support and receives this default value back, how should it render it? In english? Check if it's "Unknown location" and then replace it with a Lithuanian counterpart? What if the default value returned by the API changes?

So my suggestion is to return `null` if geocoding API can't find an address for those coordinates. This will let the API user decide how he wants to render a title that was not assigned. When we actually use this geocoder, it could be wise to log a warning if we can't find the address.

Reply:
Didn't consider the languages part at first, so I guess "Unknown location" <= string.Empty (could also technically make a check if string is empty) < null in this case